app/test/unit/shell-test.ts

```diff
@@ -0,0 +1,96 @@
+import { expect } from 'chai'
+import * as child_process from 'child_process'
+import * as os from 'os'
+import * as sinon from 'sinon'
+
+import { shellNeedsPatching, getEnvironmentFromShell } from '../../src/lib/shell'
+
+function setGlobalsForWin32() {
+  const g: any = global
+  g['__WIN32__'] = true
+  g['__DARWIN__'] = false
+}
+
+function setGlobalsForDarwin() {
+  const g: any = global
+  g['__WIN32__'] = false
+  g['__DARWIN__'] = true
+}
+
+function restoreGlobals() {
+  const g: any = global
+  g['__WIN32__'] = process.platform === 'win32'
+  g['__DARWIN__'] = process.platform === 'darwin'
+}
+
+describe('shell', () => {
+  describe('shellNeedsPatching', () => {
+
+    const oldEnv = process.env
+    const p: any = process
+
+    beforeEach(() => {
+      p._env = oldEnv
+    })
+
+    afterEach(() => {
+      process.env = oldEnv
+      delete p._env
+    })
+
+    it('returns false for Windows', () => {
+      setGlobalsForWin32()
+      expect(shellNeedsPatching(process)).to.be.false
+    })
+
+    it('returns false for macOS when it has a PWD', () => {
+      setGlobalsForDarwin()
+      process.env.PWD = __dirname
+      expect(shellNeedsPatching(process)).to.be.false
+    })
+
+    it('returns true for macOS when it is missing a PWD', () => {
+      setGlobalsForDarwin()
+      delete process.env.PWD
+      expect(shellNeedsPatching(process)).to.be.true
+    })
+  })
+
+  describe('getEnvironmentFromShell', () => {
+    let spawnSync: sinon.SinonStub | null = null
+
+    beforeEach(() => {
+      // setup for macOS as this is the only interesting environment
+      setGlobalsForDarwin()
+      spawnSync = sinon.stub(child_process, 'spawnSync')
+    })
+
+    afterEach(() => {
+      restoreGlobals()
+      spawnSync!.restore()
+    })
+
+    it('parses environment text into array', () => {
+      spawnSync!.returns({
```

Comment:
@desktop/core not sure how you feel about stubbing out the Node core APIs like this for tests, so I'll call it out now before we go too far down this path.

I could probably refactor this to do it differently (like how we do [`IAppShell`](https://github.com/desktop/desktop/blob/master/app/test/test-app-shell.ts) testing, but this felt neater and less dangerous than mocking the `child_process` API (which we are using in a few places already) and mimics how the Atom team tested this.